### DISCUSSION TITLE
SyntaxError: Cannot use import statement outside a module

### DISCUSSION THREAD

**Discussion by dev-samples** [aws/aws-cdk#26864] [Q&A]
**UPDATE:**

Below are changes needed to make the sample projet working from:
https://docs.aws.amazon.com/cdk/v2/guide/serverless_example.html

**1: import -> require and  module.exports = {handler}**
Based on: 
https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/sdk-example-javascript-syntax.html

Convert to `require` and do
```
//export const handler = async (event) => {
const handler = async (event) => {
...
module.exports = {handler}
```

**2: Fix lambda handler name in MyWidgetService/lib/widget_service.ts**

When getting this error:

`2023-08-24T08:47:06.485Z	undefined	ERROR	Uncaught Exception 	{"errorType":"Runtime.HandlerNotFound","errorMessage":"widgets.main is undefined or not exported","stack":["Runtime.HandlerNotFound: widgets.main is undefined or not exported","    at UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1096:15)","    at async start (file:///var/runtime/index.mjs:1256:23)","    at async file:///var/runtime/index.mjs:1262:1"]}
`

Change:

MyWidgetService/lib/widget_service.ts

```
    const handler = new lambda.Function(this, "WidgetHandler", {
      ...
      handler: "widgets.main",
      ...
    });

```

To:

```
    const handler = new lambda.Function(this, "WidgetHandler", {
      ...
      handler: "widgets.handler",
      ...
    });

```

**3: < HTTP/2 500 Cannot read properties of undefined (reading 'length')**

when curling:

```
❯ curl -X GET https://xxxx.execute-api.eu-west-1.amazonaws.com/prod
Cannot read properties of undefined (reading 'length')

```

From web console something else:
```

2023-08-24T09:02:09.109Z	f0a95076-d1da-4a39-956a-6a3f3ca3f3ab	ERROR	Error [UnimplementedHTTPMethodError]: Unimplemented HTTP method: undefined
    at routeRequest (/var/task/widgets.js:42:17)
    at Runtime.handler (/var/task/widgets.js:87:18)
    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1147:29)
```

which is as expected.

But would make sense if the sample projected worked out of the box.

I am trying out this example for deploying a serverless application (api gateway + lambda):
https://docs.aws.amazon.com/cdk/v2/guide/serverless_example.html

that I have deployed (**cdk deploy**) without introducing any changes to the sample files provided in above guide.

But when I test the lambda via the web console I get:

```
Test Event Name
(unsaved) test event

Response
{
  "errorType": "Runtime.UserCodeSyntaxError",
  "errorMessage": "SyntaxError: Cannot use import statement outside a module",
  "trace": [
    "Runtime.UserCodeSyntaxError: SyntaxError: Cannot use import statement outside a module",
    "    at _loadUserApp (file:///var/runtime/index.mjs:1058:17)",
    "    at async UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1093:21)",
    "    at async start (file:///var/runtime/index.mjs:1256:23)",
    "    at async file:///var/runtime/index.mjs:1262:1"
  ]
}

Function Logs
2023-08-24T07:57:50.883Z	undefined	ERROR	Uncaught Exception 	{"errorType":"Runtime.UserCodeSyntaxError","errorMessage":"SyntaxError: Cannot use import statement outside a module","stack":["Runtime.UserCodeSyntaxError: SyntaxError: Cannot use import statement outside a module","    at _loadUserApp (file:///var/runtime/index.mjs:1058:17)","    at async UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1093:21)","    at async start (file:///var/runtime/index.mjs:1256:23)","    at async file:///var/runtime/index.mjs:1262:1"]}
2023-08-24T07:57:52.482Z	undefined	ERROR	Uncaught Exception 	{"errorType":"Runtime.UserCodeSyntaxError","errorMessage":"SyntaxError: Cannot use import statement outside a module","stack":["Runtime.UserCodeSyntaxError: SyntaxError: Cannot use import statement outside a module","    at _loadUserApp (file:///var/runtime/index.mjs:1058:17)","    at async UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1093:21)","    at async start (file:///var/runtime/index.mjs:1256:23)","    at async file:///var/runtime/index.mjs:1262:1"]}
START RequestId: 448b3cbb-e043-4571-b34a-44bf48a7a697 Version: $LATEST
```

From below tsconfig.json file it seems this generated sample project is a **commonjs** project which use _require_ instead of _import_:

https://www.scaler.com/topics/nodejs/require-vs-import-nodejs/

https://stackoverflow.com/questions/76065961/syntaxerror-unexpected-token-export-in-aws-lambda-node-14-x

https://stackoverflow.com/questions/70849225/how-to-solve-cannot-use-import-statement-outside-a-module-in-aws-lambda-console

So if I change the handler **resources/widgets.js** (again copy/pasted directly from the guide above) from:

```
import { S3Client, ListObjectsCommand } from "@aws-sdk/client-s3";
...

```

To:

```
const { S3Client, ListObjectsCommand } = require("@aws-sdk/client-s3");
...
```

Deploy that and run the test again from the web console I now get:

```
{
  "errorType": "Runtime.UserCodeSyntaxError",
  "errorMessage": "SyntaxError: Unexpected token 'export'",
  "trace": [
    "Runtime.UserCodeSyntaxError: SyntaxError: Unexpected token 'export'",
    "    at _loadUserApp (file:///var/runtime/index.mjs:1058:17)",
    "    at async UserFunction.js.module.exports.load (file:///var/runtime/index.mjs:1093:21)",
    "    at async start (file:///var/runtime/index.mjs:1256:23)",
    "    at async file:///var/runtime/index.mjs:1262:1"
  ]
```

Any input?

and the generated **package.json** for the above hello world project:

```
{
  "name": "my_widget_service",
  "version": "0.1.0",
  "bin": {
    "my_widget_service": "bin/my_widget_service.js"
  },
  "scripts": {
    "build": "tsc",
    "watch": "tsc -w",
    "test": "jest",
    "cdk": "cdk"
  },
  "devDependencies": {
    "@types/jest": "^29.5.3",
    "@types/node": "20.4.5",
    "jest": "^29.6.2",
    "ts-jest": "^29.1.1",
    "aws-cdk": "2.90.0",
    "ts-node": "^10.9.1",
    "typescript": "~5.1.6"
  },
  "dependencies": {
    "aws-cdk-lib": "2.90.0",
    "constructs": "^10.0.0",
    "source-map-support": "^0.5.21"
  }
}

```

**tsconfig.json**

```
{
  "compilerOptions": {
    "target": "ES2020",
    "module": "commonjs",
    "lib": [
      "es2020",
      "dom"
    ],
    "declaration": true,
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": false,
    "inlineSourceMap": true,
    "inlineSources": true,
    "experimentalDecorators": true,
    "strictPropertyInitialization": false,
    "typeRoots": [
      "./node_modules/@types"
    ]
  },
  "exclude": [
    "node_modules",
    "cdk.out"
  ]
}

```

**Comment by iqbal86** [comment]
The error message indicates a syntax error in your code related to the use of the export keyword. Please replace it with 'exports' and let me know if that works for you!

**Comment by kevinsea** [comment]
I was doing the C# demo and ran into the same issue with the js lambda.  I'm not a JavaScript guy, especially the older versions and that lambda JavaScript is older.   What did to get the demo to work was:
- Use an AI tool to convert the old JavaScript to typescript using js2ts.com
- Compiled the ts down to modern js
- Apply fix 2 from @dev-samples above

Files:
[widgets.js.txt](https://github.com/aws/aws-cdk/files/14442233/widgets.js.txt)
[widgets.ts.txt](https://github.com/aws/aws-cdk/files/14442234/widgets.ts.txt)
